core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/CrtHttpRequestConverter.java

```diff
@@ -0,0 +1,151 @@
+
+    /**
+     * Convert an {@link SdkHttpRequest} to an {@link HttpRequest}.
+     */
+    public static HttpRequest toRequest(SdkHttpRequest request, ContentStreamProvider payload) {
+        String method = request.method().name();
+        String encodedPath = encodedPathToCrtFormat(request.encodedPath());
+
+        String encodedQueryString = request.encodedQueryParameters().map(value -> "?" + value).orElse("");
+
+        HttpHeader[] crtHeaderArray = createHttpHeaderArray(request);
+
+        HttpRequestBodyStream crtInputStream = null;
+        if (payload != null) {
+            crtInputStream = new CrtInputStream(payload, READ_BUFFER_SIZE);
+        }
+
+        return new HttpRequest(method, encodedPath + encodedQueryString, crtHeaderArray, crtInputStream);
+    }
+
+    /**
+     * Convert an {@link HttpRequest} to an {@link SdkHttpRequest}.
+     */
+    public static SdkHttpRequest toRequest(SdkHttpRequest request, HttpRequest crtRequest) {
+        SdkHttpRequest.Builder builder = request.toBuilder();
+
+        builder.clearHeaders();
+        for (HttpHeader header : crtRequest.getHeaders()) {
+            builder.appendHeader(header.getName(), header.getValue());
+        }
+
+        URI fullUri;
+        try {
+            String portString = SdkHttpUtils.isUsingStandardPort(builder.protocol(), builder.port()) ? "" : ":" + builder.port();
+            String encodedPath = encodedPathFromCrtFormat(request.encodedPath(), crtRequest.getEncodedPath());
+            String fullUriString = builder.protocol() + "://" + builder.host() + portString + encodedPath;
+            fullUri = new URI(fullUriString);
+        } catch (URISyntaxException e) {
+            return null;
+        }
+
+        builder.encodedPath(fullUri.getRawPath());
+        String remainingQuery = fullUri.getQuery();
+
+        builder.clearQueryParameters();
+        while (remainingQuery != null && remainingQuery.length() > 0) {
+            int nextQuery = remainingQuery.indexOf('&');
+            int nextAssign = remainingQuery.indexOf('=');
+            if (nextAssign < nextQuery || (nextAssign >= 0 && nextQuery < 0)) {
+                String queryName = remainingQuery.substring(0, nextAssign);
+                String queryValue = remainingQuery.substring(nextAssign + 1);
+                if (nextQuery >= 0) {
+                    queryValue = remainingQuery.substring(nextAssign + 1, nextQuery);
+                }
+
+                builder.appendRawQueryParameter(queryName, queryValue);
+            } else {
+                String queryName = remainingQuery;
+                if (nextQuery >= 0) {
+                    queryName = remainingQuery.substring(0, nextQuery);
+                }
+
+                builder.appendRawQueryParameter(queryName, null);
+            }
+
+            if (nextQuery >= 0) {
+                remainingQuery = remainingQuery.substring(nextQuery + 1);
+            } else {
+                break;
+            }
+        }
```

Comment:
This scares me. Can I assume it's just copied and we don't need to re-review it?